MsCorePkg/Library/MuSecureBootKeySelectorLib/MuSecureBootKeySelectorLib.c

```diff
@@ -17,9 +17,228 @@
+    //
+    // Create an instance of the hash protocol for this controller.
+    //
+    Status = Hash2ServiceBinding->CreateChild (Hash2ServiceBinding, &mHash2ServiceHandle);
+    if (EFI_ERROR (Status)) {
+      Status = EFI_UNSUPPORTED;
+      goto Exit;
+    }
+  }
+
+  //
+  // Check if the current db matches a Microsoft Serviced Key.
+  // walk each entry in the DbVar and check if it matches the MS Serviced Key.
+  //
+  BOOLEAN  IsMsftCertificate = FALSE;
+
+  CertList = (EFI_SIGNATURE_LIST *)DbVar;
+  while ((DbVarSize > 0) && (DbVarSize >= (UINTN)CertList->SignatureListSize)) {
+    //
+    // Make sure the signature type is X.509.
+    //
+    if (CompareGuid (&CertList->SignatureType, &gEfiCertX509Guid)) {
+      Cert      = (EFI_SIGNATURE_DATA *)((UINT8 *)CertList + sizeof (EFI_SIGNATURE_LIST) + CertList->SignatureHeaderSize);
+      CertCount = (CertList->SignatureListSize - sizeof (EFI_SIGNATURE_LIST) - CertList->SignatureHeaderSize) / CertList->SignatureSize;
+
+      IsMsftCertificate = FALSE;
+      for (Index = 0; Index < CertCount; Index++) {
+        //
+        // Iterate each Signature Data Node within this CertList for verify.
+        //
+        RootCert     = Cert->SignatureData;
+        RootCertSize = CertList->SignatureSize - sizeof (EFI_GUID);
+        Status       = Hash2Protocol->HashInit (Hash2Protocol, &gEfiHashAlgorithmSha256Guid);
+        if (EFI_ERROR (Status)) {
+          goto Exit;
+        }
+
+        Status = Hash2Protocol->HashUpdate (Hash2Protocol, RootCert, RootCertSize);
+        if (EFI_ERROR (Status)) {
+          goto Exit;
+        }
+
+        Status = Hash2Protocol->HashFinal (Hash2Protocol, &HashOutput);
+        if (EFI_ERROR (Status)) {
+          goto Exit;
+        }
+
+        //
+        // Compare the hash value with the MSFT certificate hashes.
+        //
+        for (CertIndex = 0; CertIndex < sizeof (mMicrosoftCertificates) / sizeof (mMicrosoftCertificates[0]); CertIndex++) {
+          if (CompareMem (HashOutput.Sha256Hash, mMicrosoftCertificates[CertIndex], sizeof (HashOutput.Sha256Hash)) == 0) {
+            //
+            // Found a match, this is a Microsoft certificate
+            //
+            IsMsftCertificate = TRUE;
+            break;
+          }
+        }
+
+        //
+        // If no match was found, the machine is "CUSTOM"
+        //
+        if (!IsMsftCertificate) {
+          DEBUG ((DEBUG_ERROR, "[%a] - Certificate does not match any MSFT certificates. Machine is CUSTOM.\n", __FUNCTION__));
+          Status = EFI_NOT_FOUND;
+          goto Exit;
+        }
+
+        Cert = (EFI_SIGNATURE_DATA *)((UINT8 *)Cert + CertList->SignatureSize);
+      }
+    }
+
+    DbVarSize -= CertList->SignatureListSize;
+    CertList   = (EFI_SIGNATURE_LIST *)((UINT8 *)CertList + CertList->SignatureListSize);
+  }
+
+  //
+  // If we get here than we can assume that the current db matches some configuration of MS Serviced Keys.
+  //
+  if (IsMsServicedConfig != NULL) {

```

Comment:
Should the logic around `IsMsftCertificate` be reworked to factor in here? Could passing a `DbVarSize` of `0` cause this condition to hit (skipping the loop)? Is that expected?